render/cross_section/cross_section_rendering_engine_4d.cpp

```diff
@@ -0,0 +1,133 @@
+	RS::get_singleton()->instance_set_ignore_culling(instance, true);
+	Ref<World3D> world = viewport.find_world_3d();
+	if (world.is_valid()) {
+		RS::get_singleton()->instance_set_scenario(instance, world->get_scenario());
+	}
+
+	return instance;
+}
+
+void CrossSectionRenderingEngine4D::render_frame() {
+	ERR_FAIL_NULL(RenderingServer::get_singleton());
+	ERR_FAIL_NULL(get_camera());
+	ERR_FAIL_NULL(get_viewport());
+	update_camera();
+	// Maps global to cameral-local, aka world space to view space.
+	TypedArray<MeshInstance4D> mesh_instances = get_mesh_instances();
+	TypedArray<Projection> modelview_basises = get_mesh_relative_basises();
+	PackedVector4Array modelview_origins = get_mesh_relative_positions();
+	size_t instance_index = 0;
+	for (int mesh_index = 0; mesh_index < mesh_instances.size(); mesh_index++) {
+		MeshInstance4D *mesh_instance = Object::cast_to<MeshInstance4D>(mesh_instances[mesh_index]);
+		ERR_CONTINUE(mesh_instance == nullptr);
+
+		Ref<Mesh4D> mesh = mesh_instance->get_mesh();
+		if (mesh.is_null()) {
+			continue;
+		}
+		Ref<Mesh> mesh_3d = mesh->get_cross_section_mesh();
+		ERR_CONTINUE(mesh_3d.is_null());
+
+		while (_instances_3d.size() <= instance_index) {
+			_instances_3d.push_back(create_instance(*get_viewport()));
+		}
+		RID instance_3d = _instances_3d[instance_index];
+
+		RS::get_singleton()->instance_set_base(instance_3d, mesh_3d->get_rid());
+
+		Ref<Material4D> override_material = mesh_instance->get_material_override();
+		if (override_material.is_valid()) {
+			Ref<Material> override_material_3d = override_material->get_cross_section_material();
+			ERR_CONTINUE(override_material_3d.is_null());
+			RS::get_singleton()->instance_set_surface_override_material(instance_3d, 0, override_material_3d->get_rid());
+		}
+
+		Projection modelview_basis = modelview_basises[mesh_index];
+		Vector4 modelview_origin = modelview_origins[mesh_index];
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_origin", modelview_origin);
+		// Can't pass a mat4 through instance uniforms, need to break up into columns.
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_x", modelview_basis.columns[0]);
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_y", modelview_basis.columns[1]);
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_z", modelview_basis.columns[2]);
+		RS::get_singleton()->instance_geometry_set_shader_parameter(instance_3d, "modelview_basis_w", modelview_basis.columns[3]);
+
+		instance_index++;
+	}
+	while (_instances_3d.size() > instance_index) {
+		RS::get_singleton()->free(_instances_3d.pop_back());
+	}
+}
+
+void CrossSectionRenderingEngine4D::update_camera() {
+	// TODO 3D preview is broken, mesh is not drawn relative to preview camera.
+	ERR_FAIL_NULL(RenderingServer::get_singleton());
+	if (_cross_section_camera.is_null()) {
```

Comment:
`RID.is_null` is missing from the API (https://github.com/godotengine/godot/pull/107141) so we need to use `!` and `RID.is_valid()` instead as a workaround:

```suggestion
	if (!_cross_section_camera.is_valid()) {
```